### PR TITLE
Refactor AssemblyNamesTypeResolutionService to replace HashTable

### PR DESCRIPTION
**Pull request by elachlan**
Replaced the Hashtables, enabled nullability, removed `AssemblyName.CodeBase` call.

Related: #8143, #7308

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8213)

### CODE REVIEW

**Comment by elachlan** [comment]
@dreddy-work I think in this case, the fallback on Assembly.Load() failure looks into the names array for a matching entry and then attempts to load from the name located there.

**Comment by elachlan** [comment]
@dreddy-work Flaky tests?

**Comment by dreddy-work** [comment]
> @dreddy-work Flaky tests?

@elachlan , seems so. reset it for now

File: src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs [hunk 1]
```diff
@@ -78,31 +65,26 @@ public Assembly GetAssembly(AssemblyName name, bool throwOnError)
         [UnconditionalSuppressMessage("SingleFile", "IL3002", Justification = "Returns null if in a single file")]
         public string GetPathOfAssembly(AssemblyName name)
         {
-#pragma warning disable SYSLIB0044 // Type or member is obsolete. Ref https://github.com/dotnet/winforms/issues/7308
-            return name.CodeBase;
-#pragma warning restore SYSLIB0044 // Type or member is obsolete
+            try
+            {
+                return Assembly.Load(name).Location;
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs:70]: Public API earlier was not loading assembly but now it does. I do not know repercussions here.  @JeremyKuhne / @Tanya-Solyanik  any inputs/comments on this change?

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs:70]: Change is helping remove obsolete API.

**Comment by dreddy-work** [comment]
> @dreddy-work I think in this case, the fallback on Assembly.Load() failure looks into the names array for a matching entry and then attempts to load from the name located there.

My only concern is the change in public API now loads assembly earlier not.

File: src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs (hunk 1 above)
Comment (elachlan) [reply src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs:70]: Yes, its an internal class.

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs:70]: Assembly.Load and `LoadFrom` use different load context. In this call hierarchy, direct load failed and ending up calling with `LoadFrom.`  Accessing Location this way throw exception at best.  @elachlan did you run any scenario here?

Comment (elachlan) [reply src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs:70]: I did not. I'd love to find an alternative. I couldn't find a better one. Happy to leave it as it was originally, as it wasn't the main focus of the PR.

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs:70]: Agree. Lets deal with obsolete APIs separately.

Comment (elachlan) [reply src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs]: Done.

**Review by dreddy-work — APPROVED** [review]

File: src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs [hunk 2]
```diff
@@ -196,13 +173,10 @@ public Type GetType(string name, bool throwOnError, bool ignoreCase)
         ///  This is matching %windir%\Microsoft.NET\Framework*, so both 32bit and 64bit framework will be covered.
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs:173]: @elachlan  - could you please fix this comment in a follow up PR?

File: src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs (hunk 1 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs]: @elachlan . @dreddy-work  - I don't think that this use case of CodeBase is obsoleted. Docs say, don't load assembly from the Codebase path and `Do not use an AssemblyName with only the CodeBase property set.`. The harm from a new Assembly.Load is higher than that from a suppressed warning.

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Resources/AssemblyNamesTypeResolutionService.cs]: agree and for that reason, we reverted that change.

**Pull request merged**